src/main.cpp

```diff
@@ -1760,14 +1761,25 @@ CAmount GetBlockSubsidy(int nHeight, const Consensus::Params& consensusParams)
     }
 
     assert(nHeight > consensusParams.SubsidySlowStartShift());
-    int halvings = (nHeight - consensusParams.SubsidySlowStartShift()) / consensusParams.nSubsidyHalvingInterval;
+
+    int halvings = consensusParams.Halvings(nHeight);
+
     // Force block reward to zero when right shift is undefined.
     if (halvings >= 64)
         return 0;
 
-    // Subsidy is cut in half every 840,000 blocks which will occur approximately every 4 years.
-    nSubsidy >>= halvings;
-    return nSubsidy;
+    // zip208
+    // BlockSubsidy(height) :=
+    // SlowStartRate · height, if height < SlowStartInterval / 2
+    // SlowStartRate · (height + 1), if SlowStartInterval / 2 ≤ height and height < SlowStartInterval
+    // floor(MaxBlockSubsidy / 2^Halving(height)), if SlowStartInterval ≤ height and not IsBlossomActivated(height)
+    // floor(MaxBlockSubsidy / (BlossomPoWTargetSpacingRatio · 2^Halving(height))), otherwise
+    if (consensusParams.NetworkUpgradeActive(nHeight, Consensus::UPGRADE_BLOSSOM)) {
+        return (nSubsidy / Consensus::BLOSSOM_POW_TARGET_SPACING_RATIO) >> halvings;
```

Comment:
The right shift operator is only defined if `halvings` is [less than the length in bits of the promoted left operand](https://www.viva64.com/en/w/v610/), i.e. less than 64.

```suggestion
        return halvings > 63 ? 0 : ((nSubsidy / Consensus::BLOSSOM_POW_TARGET_SPACING_RATIO) >> halvings);
```